minetest/minetest
# Add particle blend mode "clip" (2nd try)

Second attempt at #14353. Now rebased on top of the batching work.

This adds a "feature", which is really a fix for #3761: `blend = "clip"` now allows particles which do not need alpha blending to **render correctly** [^1]. In particular, this means that they can be visible through translucent nodes such as water (or objects, or translucent particles).

I think there are many good reasons for this "bugfix feature":

1. It allows for correct rendering of particles which do not need translucency, which many of them are (or can be made to be).
2. A Luanti game developer has explicitly requested it.
3. It would make particles support a blend mode nodes and objects already have. [^2]
4. Correct translucency will come at a cost, which "clip" particles need not incur.

## To do

This PR is Ready for Review.

## How to test

Spawn the "clip" test particle by right-clicking with the particlespawner testtool. Observe that it is visible through water:

![Screenshot from 2024-02-05 19-19-40](https://github.com/minetest/minetest/assets/34514239/7047ea82-1140-41bf-b6d4-47df73c8f08c)

Also dig some `use_texture_alpha = "clip"` nodes which are underwater (your camera should not be underwater). The dig particles should now be visible.

(In contrast to the "blend" test particle, which is not visible through water.)

Also check that backwards compatibility works (particles should just use "blend" on older clients).

[^1]: Irrlicht renders solid scene nodes in a render pass before transparent scene nodes & "effects", as it should.
[^2]: If a "clip" blend mode makes sense for objects, it also makes sense for particles.
